Fix issue 14433 - std.encoding does not support Central European ISO-8859-2 and CP1250 encodings

chalucha opened the pull request:
This is my first pull so I hope its done right ;-)

I just added support for both encodings in a same style as current encodings are.
I came across some spacing issues so fixed them too and can't figure what was `cast(string)` for in:
```d
return
[
    cast(string)
    "CP819",
    "IBM819",
    "ISO-8859-1"
]
```
so removed it too.

[Issue 14433](https://issues.dlang.org/show_bug.cgi?id=14433)

Orvid commented on std/encoding.d:
```diff
@@ -796,6 +807,235 @@ template EncoderInstance(CharType : Latin1Char)
+Defines an Latin2-encoded string (as an array of $(D
+immutable(Latin2Char))).
+ */
+alias Latin2String = immutable(Latin2Char)[]; ///
```
What's the extra comment at the end of this for?

Orvid commented on std/encoding.d:
```diff
@@ -1,4 +1,4 @@
-// Written in the D programming language.
+﻿// Written in the D programming language.
```
Did we mess with the byte order marker here?

Orvid commented on std/encoding.d:
```diff
@@ -796,6 +807,235 @@ template EncoderInstance(CharType : Latin1Char)
+    {
+        if (c < 0x80) return true;
+        if (c >= 0xFFFD) return false;
+        foreach(wchar d;charMap) { if (c == d) return true; }
```
Yikes.... Are the extra characters that the encoding supports in any sets of continuous  sets? If so, we should be doing a set of branches here, as this is currently far from optimal.

yebblies replied:
Wouldn't it be nice if we could do it with an immutable AA?

chalucha replied:
Hi, I've got finally back to this.
I am not sure if AAs are the best way to solve this? My plan was to use AA for canEncode and encode methods and current char array for decode methods.

I tried some simple testing with AAs (http://pastebin.com/U1yvfPLM)
With these results (all release with dub):
DMD (2.067):
Array:  1 sec, 764 ms, 912 μs, and 1 hnsec
AA:     294 ms, 69 μs, and 2 hnsecs

LDC2 (0.15.1):
Array:  1 hnsec
AA:     253 ms, 652 μs, and 1 hnsec

GDC (4.8.4):
Array:  0 hnsecs
AA:     238 ms, 871 μs, and 5 hnsecs

I guess that the problem is with current AA implementation so implement using AA and hope for the future AA reimplementation? Thx for advice.

chalucha replied:
I've made sorted list of characters over 0x80 (for CP1250) to see if there are some continuous sets here: http://pastebin.com/Fu7n9yU6
There are some, but with a lot of holes.

Orvid commented on std/encoding.d:
```diff
@@ -796,6 +807,235 @@ template EncoderInstance(CharType : Latin1Char)
+        else
+        {
+            ptrdiff_t n = -1;
+            foreach (i, wchar d; charMap)
```
See comment above.

Orvid commented on std/encoding.d:
```diff
@@ -796,6 +807,235 @@ template EncoderInstance(CharType : Latin1Char)
+        else
+        {
+            ptrdiff_t n = -1;
+            foreach (i, wchar d; charMap)
```
See comment above.

Orvid commented on std/encoding.d:
```diff
@@ -796,6 +807,235 @@ template EncoderInstance(CharType : Latin1Char)
+    {
+        if (c < 0xA0) return true;
+        if (c >= 0xFFFD) return false;
+        foreach(wchar d;charMap) { if (c == d) return true; }
```
See comment below.

quickfur replied:
Style nitpick: please split the `foreach` into multiple lines for better readability:
````
foreach (wchar d; charMap)
{
    if (c == d) return true;
}
````

Orvid commented:
For future reference, it's easier to review if you put the formatting changes in a separate commit.

chalucha commented:
Comment tags was copy pasted so I removed them from all encodings.
About BOM - I checked other phobos files and they are mix of ASCII and UTF8 without BOM so I removed it too.
Maybe it should be mentioned here? http://wiki.dlang.org/Pull_Requests

chalucha commented:
I've baked better tests with more possible solutions here: http://pastebin.com/C6jDYzw0

Algorithms used:
* Array - currently used slow method - O(n)
* AA - use of associative array as a map
* IF - simple and ugly generated chain of IFs
* RBT - use Red Black Tree as character map
* BS - use simple binary search over sorted array
* BST - use binary search tree encoded to immutable array

I think that AA is the most elegant one, but currently not as performant as RBT or BST.
I would preffer BST array search because of its simplicity.

So please advice which way to go, I'll then modify current pull and create the other one to change current encodings the same way.

Results on my linux machine are:
#### DMD

Algorithm | Duration
------------ | -----------
Array | 4 secs, 144 ms, 239 μs, and 6 hnsecs
AA | 1 sec, 408 ms, 367 μs, and 4 hnsecs
IF | 1 sec, 113 ms, and 779 μs
RBT | 1 sec, 320 ms, 87 μs, and 3 hnsecs
BS | 1 sec, 444 ms, 894 μs, and 7 hnsecs
BST | 1 sec, 295 ms, 103 μs, and 1 hnsec

#### GDC

Algorithm | Duration
------------ | -----------
Array | 4 secs, 415 ms, 200 μs, and 6 hnsecs
AA | 848 ms, 581 μs, and 3 hnsecs
IF | 843 ms and 503 μs
RBT | 398 ms and 407 μs
BS | 739 ms, 749 μs, and 5 hnsecs
BST | 422 ms, 399 μs, and 6 hnsecs

#### LDC2

Algorithm | Duration
------------ | -----------
Array | 2 secs, 176 ms, 820 μs, and 8 hnsecs
AA | 771 ms, 346 μs, and 1 hnsec
IF | 535 ms and 60 μs
RBT | 474 ms and 487 μs
BS | 400 ms, 19 μs, and 9 hnsecs
BST | 524 ms, 759 μs, and 8 hnsecs

quickfur commented on std/encoding.d:
```diff
@@ -796,6 +807,235 @@ template EncoderInstance(CharType : Latin1Char)
+
+/** Defines an Latin2-encoded character. */
+enum Latin2Char : ubyte { init }
+/**
```
Style nitpick: please leave a blank line before the `/**`.

quickfur commented:
I think BST is probably the way to go: reasonable performance, and reasonably simple implementation. If performance becomes a problem in the future, then we can consider how to improve it. Let's not get stuck on performance, because the functionality is more important right now.

chalucha commented on std/range/package.d:
```diff
@@ -8665,10 +8665,11 @@ struct NullSink
 {
     import std.algorithm : map, copy;
-    [4, 5, 6].map!(x => x * 2).copy(NullSink());
+    [4, 5, 6].map!(x => x * 2).copy(NullSink()); // data is discarded
```
Actually I made git pull --rebase before further modifications and I did not touch this file. Its even same as previous version in my fork - anyone knows how to fix this is PR?

chalucha commented:
I've updated the PR with in array binary search tree implementation to find right value for dchar. Other way around stays the same O(1).
I used simple generation utility from unicode mappings spec files to Encoder template so it would be easier to change it as required and safer for human errors.
After this gets accepted, I'll try to speed up the original encodings the same way too and even add some other encodings with it.

Sources for the utility: https://github.com/chalucha/encgen

DmitryOlshansky commented on std/encoding.d:
```diff
@@ -796,6 +807,333 @@ template EncoderInstance(CharType : Latin1Char)
+ */
+alias Latin2String = immutable(Latin2Char)[];
+
+template EncoderInstance(CharType : Latin2Char)
```
deafult visibility is public, please check that internals are private

DmitryOlshansky commented:
You might have done `git pull --rebase` against your own branch. Correct way would be `git pull --rebase upstream master`.

This seems solid but std.encoding is generally considered crappy and up for replacement. 

Anyway I'd merge it if:
 - there are no merge commits and other crap
 - most symobls are made private unless meant for the end-user

chalucha commented:
Remade it in #3626 for clean up..
